datafusion/physical-plan/src/sorts/sort.rs

```diff
@@ -737,9 +738,30 @@ impl SortExec {
     /// This can reduce the memory pressure required by the sort
     /// operation since rows that are not going to be included
     /// can be dropped.
-    pub fn with_fetch(mut self, fetch: Option<usize>) -> Self {
-        self.fetch = fetch;
-        self
+    pub fn with_fetch(&self, fetch: Option<usize>) -> Self {
+        let mut cache = self.cache.clone();
+        if fetch.is_some() {
+            // When a theoretically unnecessary sort becomes a top-K (which
```

Comment:
I don't fully understand how a top-k sort would become bounded. I may misundersrtand what the [ExecutionMode](https://docs.rs/datafusion/latest/datafusion/physical_plan/enum.ExecutionMode.html) trait means, but it seems like `TopK` could not complete until its input completed, but if its input was unbounded the sort itself therefore would also be unbounded

Reply:
In general you are right -- however, you are missing that we turn it into `Bounded` only when the sort requirement is already satisfied. This happens when a sort "becomes" unnecessary during one of the plan optimization steps (and it will eventually get removed).

Reply:
> I don't fully understand how a top-k sort would become bounded. I may misundersrtand what the [ExecutionMode](https://docs.rs/datafusion/latest/datafusion/physical_plan/enum.ExecutionMode.html) trait means, but it seems like `TopK` could not complete until its input completed, but if its input was unbounded the sort itself therefore would also be unbounded

I misassumed the implementation of top-k. Could you please take a look to the new idea?

Reply:
FWIW I think the comments help explain what is going on here well. Thank you